src/parsing.jl

```diff
@@ -272,12 +292,23 @@ function pqparse(::Type{ZonedDateTime}, str::AbstractString)
         return ZonedDateTime(typemin(DateTime), tz"UTC")
     end
 
-    for fmt in TIMESTAMPTZ_FORMATS[1:(end - 1)]
-        parsed = tryparse(ZonedDateTime, str, fmt)
-        parsed !== nothing && return parsed
+    return _pqparse(ZonedDateTime, str)
+end
+
+function pqparse(::Type{UTCDateTime}, str::AbstractString)
+    if str == "infinity"
+        depwarn_timetype_inf()
+        return UTCDateTime(typemax(DateTime))
+    elseif str == "-infinity"
+        depwarn_timetype_inf()
+        return UTCDateTime(typemin(DateTime))
     end
 
-    return parse(ZonedDateTime, _trunc_seconds(str), TIMESTAMPTZ_FORMATS[end])
+    # Postgres should give us strings ending with +00, +00:00, -00:00
+    # We use the regex below to strip these character off before parsing, iff,
+    # the values after the `-`/`+` are `0` or `:`. This means parsing will fail if
+    # we're asked to parse a non-UTC string like +04:00.
+    return _pqparse(UTCDateTime, replace(str, r"[-|\+][0|:]*$" => ""))
```

Comment:
Have you seen a scenario when `+00:00` is generated or are you assuming the need to support that based on ISO format?

For comparison, the code as committed:
```julia
rerep(str) = replace(str, r"[-|\+][0|:]*$" => "")
```
```
julia> @benchmark rerep("2022-01-01 00:00:00+00")
BenchmarkTools.Trial: 10000 samples with 199 evaluations.
 Range (min … max):  424.603 ns …   8.496 μs  ┊ GC (min … max): 0.00% … 94.26%
 Time  (median):     434.817 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   533.724 ns ± 288.034 ns  ┊ GC (mean ± σ):  1.36% ±  2.95%

  █▅▄▂▂▁      ▃▂▁                              ▁▁▁▂▁▁           ▁
  ██████████▇████▇▇▇▇█▇▇▇▆▇▇▇▇█▇▇▇▇▇▇█▇▇▇██▇██████████▇▆▆▅▅▅▅▅▅ █
  425 ns        Histogram: log(frequency) by time       1.06 μs <

 Memory estimate: 240 bytes, allocs estimate: 4.
```
(not sure, but I believe allocations within PCRE2 are hidden from Julia)

Assuming we always get `+00`, we can do this instead:

```julia
strrep(str) = replace(str, "+00" => "")
```
```
julia> @benchmark strrep("2022-01-01 00:00:00+00")
BenchmarkTools.Trial: 10000 samples with 331 evaluations.
 Range (min … max):  254.018 ns …   6.491 μs  ┊ GC (min … max): 0.00% … 94.32%
 Time  (median):     264.082 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   328.889 ns ± 249.978 ns  ┊ GC (mean ± σ):  3.93% ±  5.41%

  ██▄▄▃▃▃▁▁▁▁▁ ▁                     ▁▁ ▁▁▁▁▁▁                  ▂
  █████████████████████▇▇██▇█████▇██████████████▇█▇▇█▇▇▇▇▇▆▄▄▅▅ █
  254 ns        Histogram: log(frequency) by time        699 ns <

 Memory estimate: 496 bytes, allocs estimate: 8.
```

And if you just assume that it's always going to be `+` and not `-`, you can do a less-allocating version:
```julia
function ssrep(str)
    pos = findlast('+', str)
    return SubString(pstr, 1, prevind(pstr, pos))
end
```
```
julia> @benchmark ssrep("2022-01-01 00:00:00+00")
BenchmarkTools.Trial: 10000 samples with 442 evaluations.
 Range (min … max):  233.887 ns …   4.970 μs  ┊ GC (min … max): 0.00% … 92.22%
 Time  (median):     241.057 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   275.007 ns ± 107.643 ns  ┊ GC (mean ± σ):  0.42% ±  1.58%

  ██▅▃▃▁▁▁▁▁▁▁                                                  ▁
  ███████████████▇▇█▇█▇▇▇▇▇▇▇▇▇▇▇▆▇▇▇▇▇▇▇▇██▇██▇▇▇▆▇▇▇▆▆▇▇▆▇▇▇▆ █
  234 ns        Histogram: log(frequency) by time        592 ns <

 Memory estimate: 32 bytes, allocs estimate: 1.
```
I checked parsing time for these and the SubString version seems to be slightly slower to parse into a UTCDateTime at almost exactly the amount it's faster to generate here. So the benefit of the SubString version over the basic string-replace version is mostly around avoiding allocations/GC.

Reply:
> Have you seen a scenario when +00:00 is generated or are you assuming the need to support that based on ISO format?

Yes, my understanding is that while LibPQ.jl defaults to `+00`, but you can configure the client to return something different? In that case it seems better to error than simply drop everything after `+` which could silently give you the wrong result and doesn't allow for `-00`. I don't have a strong opinion about whether to use a regex or not, though I'll note we already use `replace` with a regex in `_trunc_seconds`, so this feels like an over optimization. 

NOTE: Interestingly, the difference between `rerep` and `strrep` on my system is much smaller running 1.6.7 on an m1 max.

Reply:
> Yes, my understanding is that while LibPQ.jl defaults to +00, but you can configure the client to return something different?

You can do this, but many other assumptions are made in this code and I don't believe the choice to allow `-00` or `+00:00` extends to any of the other specific formats that PostgreSQL supports. Here are some examples on PostgreSQL 14 (client and server):

```
postgres=# SELECT '2022-01-02 23:15:22'::timestamptz;
      timestamptz
------------------------
 2022-01-02 23:15:22-06
(1 row)
```
This is what you get when you have a time zone that's offset. We set UTC in our connections:
```
postgres=# SET timezone = 'UTC';
SET
postgres=# SELECT '2022-01-02 23:15:22'::timestamptz;
      timestamptz
------------------------
 2022-01-02 23:15:22+00
(1 row)
```

My machine is already set to the format we expect to parse in LibPQ.jl (`ISO,YMD`), but we can see what the other formats look like (I passed `YMD` here in addition but AFAICT it only affects the `ISO` representation):
```
postgres=# SET datestyle = 'Postgres,YMD';
SET
postgres=# SELECT '2022-01-02 23:15:22'::timestamptz;
         timestamptz
------------------------------
 Sun Jan 02 23:15:22 2022 UTC
(1 row)

postgres=# SET datestyle = 'SQL,YMD';
SET
postgres=# SELECT '2022-01-02 23:15:22'::timestamptz;
       timestamptz
-------------------------
 01/02/2022 23:15:22 UTC
(1 row)

postgres=# SET datestyle = 'German,YMD';
SET
postgres=# SELECT '2022-01-02 23:15:22'::timestamptz;
       timestamptz
-------------------------
 02.01.2022 23:15:22 UTC
(1 row)
```

As far as I can tell there is no way to get an offset of `00` to print in any way other than `+00`.

Reply:
Okay, yeah, I also couldn't get it to print something different for UTC, so I'm fine to got with `strrep` for now. Without running a separate query to check what the timezone is set to I think `ssrep` is to dangerous.